libafl_qemu/src/command/mod.rs

```diff
@@ -0,0 +1,638 @@
+        {
+            pub fn new() -> Self {
+                let native_parsers = Box::new(
+                    vec![$(Box::new($native_command_parser)
+                        as Box<
+                            dyn NativeCommandParser<
+                                StdCommandManager<QT, S, SM>,
+                                StdEmulatorExitHandler<SM>,
+                                QT,
+                                S,
+                            >,
+                        >),*]
+                    .into_iter(),
+                );
+
+                let mut parsers: HashMap<
+                    GuestReg,
+                    Box<dyn NativeCommandParser<Self, StdEmulatorExitHandler<SM>, QT, S>>,
+                > = HashMap::new();
+
+                for parser in native_parsers {
+                    assert!(parsers
+                        .insert(parser.command_id(), parser)
+                        .is_none(), "Trying to use native commands with the same ID");
+                }
+
+                Self {
+                    native_command_parsers: parsers,
+                }
+            }
+        }
+
+        impl<QT, S, SM> CommandManager<StdEmulatorExitHandler<SM>, QT, S> for $name<QT, S, SM>
+        where
+            QT: QemuHelperTuple<S> + StdInstrumentationFilter<S> + Debug,
+            S: State + HasExecutions,
+            S::Input: HasTargetBytes,
+            SM: IsSnapshotManager,
+        {
+            fn parse(
+                &self,
+                qemu: Qemu,
+            ) -> Result<Rc<dyn IsCommand<Self, StdEmulatorExitHandler<SM>, QT, S>>, CommandError> {
+                let arch_regs_map: &'static EnumMap<ExitArgs, Regs> = get_exit_arch_regs();
+                let cmd_id: GuestReg = qemu.read_reg::<Regs, GuestReg>(arch_regs_map[ExitArgs::Cmd])?;
+
+                let cmd_parser = self
+                    .native_command_parsers
+                    .get(&cmd_id)
+                    .ok_or(CommandError::UnknownCommand(cmd_id))?;
+                let cmd = cmd_parser.parse(qemu, arch_regs_map)?;
+
+                Ok(cmd)
+            }
+        }
+
+        impl<QT, S, SM> Debug for $name<QT, S, SM>
+        where
+            QT: QemuHelperTuple<S> + StdInstrumentationFilter<S> + Debug,
+            S: State + HasExecutions,
+            S::Input: HasTargetBytes,
+            SM: IsSnapshotManager,
+        {
+            fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+                write!(f, stringify!($name))
+            }
+        }
+    };
+}
+
+define_std_command_manager!(
+    StdCommandManager,
+    [
+        StartPhysCommandParser,
+        StartVirtCommandParser,
+        InputPhysCommandParser,
+        InputVirtCommandParser,
+        SaveCommandParser,
+        LoadCommandParser,
+        EndCommandParser,
+        VersionCommandParser,
+        VaddrFilterAllowRangeCommandParser
+    ]
+);
+
+pub trait CommandManager<E, QT, S>: Sized
```

Comment:
Related question, when would I implement my own `CommandManager`?

Reply:
It's useful to add custom or optional commands (for example to add some commands depending on a feature being turned on) or for compatibility with other command systems (we will need it at some points to make libafl_qemu compatible with kAFL/Nyx targets for example).

The old system relies on a huge match, it makes it quite inflexible and unreadable for something different from the main usecase.

This traits makes possible something [like this](https://github.com/AFLplusplus/LibAFL/blob/53a5832ac7cb06bf2941be1bf5b372c3ff17614f/libafl_qemu/src/command/mod.rs#L142-L155), which is much cleaner imho.